api/query/cache/service/search.go

```diff
@@ -0,0 +1,220 @@
+	// are currently resident in `idx`.
+	store, err := getDatastore(ctx)
+	if err != nil {
+		return &searchErr{
+			Detail:  err,
+			Message: "Failed to open Datastore",
+			Code:    http.StatusInternalServerError,
+		}
+	}
+
+	ids := make([]int64, 0, len(rq.RunIDs))
+	runs := make([]shared.TestRun, 0, len(rq.RunIDs))
+	missing := make([]shared.TestRun, 0, len(rq.RunIDs))
+	for i := range rq.RunIDs {
+		id := index.RunID(rq.RunIDs[i])
+		run, err := idx.Run(id)
+		// If getting run metadata fails, attempt write-on-read for this run.
+		if err != nil {
+			runPtr := new(shared.TestRun)
+			if err := store.Get(store.NewIDKey("TestRun", int64(id)), runPtr); err != nil {
+				return &searchErr{
+					Detail:  err,
+					Message: "Unknown test run ID " + string(id),
+					Code:    http.StatusBadRequest,
+				}
+			}
+			runPtr.ID = int64(id)
+			go idx.IngestRun(*runPtr)
+			missing = append(missing, *runPtr)
+		} else {
+			// Ensure that both `ids` and `runs` correspond to the same test runs.
+			ids = append(ids, rq.RunIDs[i])
+			runs = append(runs, run)
+		}
+	}
+
+	// Return to client `http.StatusUnprocessableEntity` immediately if any runs
+	// are missing.
+	if len(runs) == 0 && len(missing) > 0 {
+		data, err := json.Marshal(shared.SearchResponse{
+			IgnoredRuns: missing,
+		})
+		if err != nil {
+			return &searchErr{
+				Detail:  err,
+				Message: "Failed to marshal results to JSON",
+				Code:    http.StatusInternalServerError,
+			}
+		}
+		w.WriteHeader(http.StatusUnprocessableEntity)
+		w.Write(data)
+		return nil
+	}
+
+	// Prepare user query based on `ids` that are (or at least were a moment ago)
+	// resident in `idx`. In the unlikely event that a run in `ids`/`runs` is no
+	// longer in `idx`, `idx.Bind()` below will return an error.
+	q := cq.PrepareUserQuery(ids, rq.AbstractQuery.BindToRuns(runs...))
+
+	// Configure format, from request params.
+	urlQuery := r.URL.Query()
+	subtests, _ := shared.ParseBooleanParam(urlQuery, "subtests")
+	interop, _ := shared.ParseBooleanParam(urlQuery, "interop")
+	diff, _ := shared.ParseBooleanParam(urlQuery, "diff")
+	diffFilter, _, err := shared.ParseDiffFilterParams(urlQuery)
+	if err != nil {
+		return &searchErr{
+			Detail:  err,
+			Message: "Failed to parse diff filter",
+			Code:    http.StatusBadRequest,
+		}
+	}
+	opts := query.AggregationOpts{
+		IncludeSubtests:         subtests != nil && *subtests,
+		InteropFormat:           interop != nil && *interop,
+		IncludeDiff:             diff != nil && *diff,
+		DiffFilter:              diffFilter,
+		IgnoreTestHarnessResult: shared.IsFeatureEnabled(store, "ignoreHarnessInTotal"),
+	}
+	plan, err := idx.Bind(runs, q)
+	if err != nil {
+		return &searchErr{
+			Detail:  err,
+			Message: "Failed to create query plan",
+			Code:    http.StatusInternalServerError,
+		}
+	}
+
+	results := plan.Execute(runs, opts)
+	res, ok := results.([]shared.SearchResult)
+	if !ok {
+		return &searchErr{
```

Comment:
Nice catch realizing that we used to log the wrong error (!), but in general it seems concerning that `plan.Execute` could return either nil or not a `[]shared.SearchResult` and we would get no details about that :/.

Reply:
Yeah, this is not good, but it should be fixed in another PR (and probably when we run into real problems).